arrow-fx/src/test/kotlin/arrow/fx/QueueTest.kt

```diff
@@ -181,7 +385,161 @@ class QueueTest : UnitSpec() {
+          !effect { received shouldBe None }
+          !effect { (elapsed >= 100) shouldBe true }
+        }.equalUnderTheLaw(IO.unit)
+      }
+
+      "$label - queue cannot be filled at once without enough capacity" {
+        forAll(Gen.nonEmptyList(Gen.int())) { l ->
+          IO.fx {
+            val q = !queue(l.size)
+            val succeed = !q.tryOfferAll(l.toList() + 1)
+            val res = !q.takeAll()
+            Tuple2(succeed, res)
+          }.equalUnderTheLaw(IO.just(Tuple2(false, emptyList<Int>())))
+        }
+      }
+
+      "$label - can offerAll at capacity with take" {
+        IO.fx {
+          val q = !queue(1)
+          val (join, _) = !q.take().fork()
+          !IO.sleep(250.milliseconds)
+          !q.offerAll(1, 2)
+          val a = !q.take()
+          val b = !join
+          setOf(a, b)
+        }.equalUnderTheLaw(IO.just(setOf(1, 2)))
+      }
+
+      "$label - can tryOfferAll at capacity with take" {
+        IO.fx {
+          val q = !queue(1)
+          val (join, _) = !q.take().fork()
+          !IO.sleep(250.milliseconds)
+          val succeed = !q.tryOfferAll(1, 2)
+          val a = !q.take()
+          val b = !join
+          Tuple2(succeed, setOf(a, b))
+        }.equalUnderTheLaw(IO.just(Tuple2(true, setOf(1, 2))))
+      }
+
+      "$label - offerAll blocks when values not  taken" {
+        IO.fx {
+          val q = !queue(1)
+          val start = !effect { System.currentTimeMillis() }
+          val res = !q.offerAll(1, 2, 3).map { Some(it) }
+            .waitFor(100.milliseconds, default = just(None))
+
+          val elapsed = !effect { System.currentTimeMillis() - start }
+          !effect { res shouldBe None }
+          !effect { (elapsed >= 100) shouldBe true }
+        }.equalUnderTheLaw(IO.unit)
+      }
+
+      // offerAll(fa).fork() + offerAll(fb).fork() <==> queue(fa + fb) OR queue(fb + fa)
+      "$label - offerAll is atomic" {
+        forAll(Gen.nonEmptyList(Gen.int()), Gen.nonEmptyList(Gen.int())) { fa, fb ->
+          IO.fx {
+            val q = !queue(fa.size + fb.size)
+            !q.offerAll(fa.toList()).fork()
+            !q.offerAll(fb.toList()).fork()
+
+            !IO.sleep(50.milliseconds)
+
+            val res = !q.takeAll()
+            res == (fa.toList() + fb.toList()) || res == (fb.toList() + fa.toList())
+          }.equalUnderTheLaw(IO.just(true))
+        }
+      }
+
+      "$label - takeAll takes outstanding  offers" {
+        forAll(50, Gen.nonEmptyList(Gen.int()).filter { it.size <= 100 }, Gen.choose(1, 100)) { l, capacity ->
+          IO.fx {
+            val q = !queue(capacity)
+            !l.parTraverse(NonEmptyList.traverse(), q::offer).fork()
+            !IO.sleep(50.milliseconds) // Give take callbacks a chance to register
+
+            val l = !q.takeAll()
+            l.toSet()
+          }.equalUnderTheLaw(IO.just(l.toList().toSet()))
+        }
+      }
+
+      "$label - peekAll takes outstanding  offers" {
```

Comment:
Makes sense, but does `peekAll` complete the offering tasks or do they still suspend? If they suspend then this is perfectly fine^^

Reply:
They still suspend since `peekAll` does not remove any items from the `Queue`.

It's alternative `takeAll` also completes the offering tasks, making the `Queue` effectively empty.